backup-stores/s3/src/main/java/io/camunda/zeebe/backup/s3/FileSetManager.java

```diff
@@ -59,24 +67,39 @@ CompletableFuture<FileSet> save(final String prefix, final NamedFileSet files) {
 
   private CompletableFuture<FileSet.FileMetadata> saveFile(
       final String prefix, final String fileName, final Path filePath) {
+
     if (shouldCompressFile(filePath)) {
       final var algorithm = config.compressionAlgorithm().orElseThrow();
-      final var compressed = compressFile(filePath, algorithm);
-      LOG.trace("Saving compressed file {}({}) in prefix {}", fileName, compressed, prefix);
-      return client
-          .putObject(
-              put -> put.bucket(config.bucketName()).key(prefix + fileName),
-              AsyncRequestBody.fromFile(compressed))
-          .thenRunAsync(() -> cleanupCompressedFile(compressed))
-          .thenApply(unused -> FileSet.FileMetadata.withCompression(algorithm));
+      return CompletableFuture.runAsync(uploadLimit::acquireUninterruptibly)
+          .thenApply(
+              (success) -> {
+                return compressFile(filePath, algorithm);
+              })
+          .thenCompose(
+              (compressedFile) -> {
+                LOG.trace(
+                    "Saving compressed file {}({}) in prefix {}", fileName, compressedFile, prefix);
+                return client
+                    .putObject(
+                        put -> put.bucket(config.bucketName()).key(prefix + fileName),
+                        AsyncRequestBody.fromFile(compressedFile))
+                    .thenRunAsync(() -> cleanupCompressedFile(compressedFile))
+                    .thenApply(unused -> FileSet.FileMetadata.withCompression(algorithm))
+                    .whenComplete((success, error) -> uploadLimit.release());
```

Comment:
:x: this needs to happen even if compressing the file fails. the `thenCompose` in line 78 only applies when the compression succeeded.

Reply:
I added the .exceptionally() to return and release the upload channel. I am just wondering if the compression fails if we should try to upload the file uncompressed or just fail normally, wdyt?

Reply:
This change is not correct unfortunately. Now we'd release the uploadLimit twice when uploading the compressed file fails.

What I meant before is that the `.whenComplete((success, error) -> uploadLimit.release()` should not be chained on the "inner" future where we try to upload the file but on the "outer" future that represents compression and upload.

So like this:

```java
  private CompletableFuture<FileSet.FileMetadata> saveFile(
      final String prefix, final String fileName, final Path filePath) {

    if (shouldCompressFile(filePath)) {
      final var algorithm = config.compressionAlgorithm().orElseThrow();
      return CompletableFuture.runAsync(uploadLimit::acquireUninterruptibly)
          .thenApply((success) -> compressFile(filePath, algorithm))
          .thenCompose(
              (compressedFile) -> {
                LOG.trace(
                    "Saving compressed file {}({}) in prefix {}", fileName, compressedFile, prefix);
                return client
                    .putObject(
                        put -> put.bucket(config.bucketName()).key(prefix + fileName),
                        AsyncRequestBody.fromFile(compressedFile))
                    .thenRunAsync(() -> cleanupCompressedFile(compressedFile))
                    .thenApply(unused -> FileSet.FileMetadata.withCompression(algorithm));
              })
          .whenComplete((success, error) -> uploadLimit.release());
    }

    return CompletableFuture.runAsync(uploadLimit::acquireUninterruptibly)
        .thenCompose(
            (nothing) -> {
              LOG.trace("Saving file {}({}) in prefix {}", fileName, filePath, prefix);
              return client
                  .putObject(
                      put -> put.bucket(config.bucketName()).key(prefix + fileName),
                      AsyncRequestBody.fromFile(filePath))
                  .thenApply(unused -> FileSet.FileMetadata.none());
            })
        .whenComplete((success, error) -> uploadLimit.release());
  }
```